bpf/init.sh

```diff
@@ -456,11 +456,17 @@ case "${MODE}" in
 		echo "#define EPHEMERAL_MIN $CILIUM_EPHEMERAL_MIN" >> $RUNDIR/globals/node_config.h
 
 		if [ "$NODE_PORT" = "true" ]; then
-			sed -i '/^#.*NATIVE_DEV_MAC.*$/d' $RUNDIR/globals/node_config.h
-			NATIVE_DEV_MAC=$(ip link show $NATIVE_DEV | grep ether | awk '{print $2}')
-			NATIVE_DEV_MAC=$(mac2array $NATIVE_DEV_MAC)
-			echo "#define NATIVE_DEV_MAC { .addr = ${NATIVE_DEV_MAC}}" >> $RUNDIR/globals/node_config.h
-
+			MAC_BY_IFINDEX_MACRO="#define NATIVE_DEV_MAC_BY_IFINDEX(IFINDEX) ({ \\
+	union macaddr mac = {.addr = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0}}; \\
+	switch (IFINDEX) { \\\\\n"
+			MAC_BY_IFINDEX_MACRO_END="	} \\
+	mac; })"
+			IDX=$(cat /sys/class/net/${NATIVE_DEV}/ifindex)
+			MAC=$(ip link show $NATIVE_DEV | grep ether | awk '{print $2}')
+			MAC=$(mac2array $MAC)
+			MAC_BY_IFINDEX_MACRO="${MAC_BY_IFINDEX_MACRO}	case ${IDX}: {union macaddr tmp = {.addr = ${MAC}}; mac=tmp;} break; \\\\\n"
+			MAC_BY_IFINDEX_MACRO="${MAC_BY_IFINDEX_MACRO}${MAC_BY_IFINDEX_MACRO_END}"
+			echo -e "${MAC_BY_IFINDEX_MACRO}" >> $RUNDIR/globals/node_config.h
```

Comment:
Nice idea, Did you check whether LLVM generates a search-tree on this construct if there are say >8 entries?

Given the number of such devices is usually not too high that would have outside world connectivity, I wonder whether we could do the same trick to overcome the u16 ifindex value problem?

We could have code-generation for two macros (or __always_inline functions), i) that is generating the following. Pseudo code for 2 devices in hostns (one with ifindex 42, another one with 147483647):

```
static __always_inline __u8 ifindex_to_dev_id(const int ifindex)
{
    switch (ifindex) {
    case 42: return 1;
    case 147483647: return 2;
    default: return 0; // invalid one
    }
}

static __always_inline __u8 dev_id_to_ifindex(const __u8 dev_id)
{
    switch (dev_id) {
    case 1: return 42;
    case 2: return 147483647;
    default: return 0; // invalid one
    }
}
```

If we set a limit of max 255 such devices that we track at the same time, they can have whatever ifindex they'd want, but we can store the tracking more efficiently in the CT with just 8 bits. Wdyt?

Reply:
> Did you check whether LLVM generates a search-tree on this construct if there are say >=8 entries?

Noup, I haven't, but I'll do it.

> Wdyt?

Yeah, sounds like a solution. Just we need to make sure that after cilium-agent restarts with a new configuration previously used devices get the same surrogate ids.

Reply:
Yep, sounds good, lets keep tracking these items somewhere then for follow-up.